### ISSUE TITLE
Test suite for `sign_aws4!`

### ISSUE THREAD

**Issue by omus** [JuliaWeb/HTTP.jl#291]
It would be good to have a test suite for this function. Amazon provides a [test suite](https://docs.aws.amazon.com/general/latest/gr/signature-v4-test-suite.html) which we could use in order to validate the behaviour of this function.

**Comment by omus** [comment]
I did a manual test and I needed to modify what headers were being generated by `sign_aws4!` but it did pass this test:

```julia
using Base.Test, HTTP.AWS4AuthRequest, HTTP.URIs, HTTP.Headers

# Modify internals of `sign_aws4!` so that only these headers are set
# setkv(headers, "x-amz-date",  datetime)
# setkv(headers, "host",  url.host)

# get-vanilla-query-order-key-case
result = HTTP.AWS4AuthRequest.sign_aws4!(
    "GET", 
    URI("https://example.amazonaws.com/?Param2=value2&Param1=value1"), 
    Headers([]), 
    b""; 
    t=DateTime(2015,8,30,12,36),
    aws_service="service",
    aws_region="us-east-1",
    aws_secret_access_key="wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY", 
    aws_access_key_id="AKIDEXAMPLE",
)

expected = "Authorization" => "AWS4-HMAC-SHA256 Credential=AKIDEXAMPLE/20150830/us-east-1/service/aws4_request, SignedHeaders=host;x-amz-date, Signature=b97d918cfa904a5beff61c982a1b6f458b799221646efd99d3219ec94cdf2500"

@test result == expected
```

I'll also note that `sign_aws4!` did not pass the "get-header-key-duplicate" test.

**Comment by ararslan** [comment]
Something to note is that we send more headers (specifically `x-amz-content-sha256` and `Content-MD5`) than Amazon includes in their examples, so we can't necessarily compare our output to theirs 1-1 unless we rethink the headers we want to send. Technically the content hashes are sort of redundant, since the SHA-256 of the content is used in constructing the signature itself, and the request is denied if the sent signature doesn't match the signature that Amazon calculates upon receiving the request.

**Comment by omus** [comment]
I think either we could make sure that our headers are a superset of the ones in the test or we can reevaluate why we have the additional headers.

**Comment by quinnj** [comment]
It's also only required for PUT S3 requests; GET requests don't need it.

**Comment by ararslan** [comment]
Closed by the addition of tests in #357, #359, #364, #366, #368, and #369.
